contracts/staking/fuse/utils/StakingRewardsPerEpoch.sol

```diff
@@ -0,0 +1,201 @@
+		uint256 reward;
+		uint256 balance;
+		uint256 pendingStake;
+		uint256 indexOfLastEpochStaked;
+	}
+
+	event Staked(address indexed staker, uint256 amount);
+	event Withdrawn(address indexed staker, uint256 amount);
+	event RewardAdded(uint256 reward);
+	event RewardPaid(address indexed user, uint256 reward);
+	event RewardsDurationUpdated(uint256 newDuration);
+	event Recovered(address token, uint256 amount);
+
+	bytes32 public constant GUARDIAN_ROLE = keccak256("GUARDIAN_ROLE");
+	uint256 public constant PRECISION = 1e18;
+
+	IERC20 public rewardsToken;
+	IERC20 public stakingToken;
+
+	mapping(address => StakeInfo) public stakersInfo;
+
+	uint256 public lastEpochIndex;
+	uint256 public pendingStakes;
+	uint256[] public rewardsPerTokenAt;
+
+	uint256 internal _totalSupply;
+
+	modifier updateReward(address account) {
+		_updateReward(account);
+		_;
+	}
+
+	constructor(address _rewardsToken, address _stakingToken) {
+		_setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+		_setupRole(GUARDIAN_ROLE, msg.sender);
+		rewardsToken = IERC20(_rewardsToken);
+		stakingToken = IERC20(_stakingToken);
+	}
+
+	function totalSupply() external view returns (uint256) {
+		return _totalSupply;
+	}
+
+	function balanceOf(address account) external view returns (uint256) {
+		return stakersInfo[account].balance;
+	}
+
+	function stake(uint256 amount)
+		external
+		payable
+		nonReentrant
+		whenNotPaused
+		updateReward(msg.sender)
+	{
+		require(amount > 0, "Cannot stake 0");
+		_stake(msg.sender, amount);
+	}
+
+	function withdraw(uint256 _amount)
+		public
+		nonReentrant
+		updateReward(msg.sender)
+	{
+		require(_amount > 0, "Cannot withdraw 0");
+		_withdraw(msg.sender, _amount);
+	}
+
+	function getReward() public nonReentrant updateReward(msg.sender) {
+		_getReward(msg.sender);
+	}
+
+	function exit() external {
+		withdraw(stakersInfo[msg.sender].balance);
+		getReward();
+	}
+
+	function rewardPerToken() public view returns (uint256) {
+		if (_totalSupply == 0) {
+			return rewardPerTokenAt[0];
+		}
+		return
+			rewardPerTokenStored +
+			((lastTimeRewardApplicable() - lastUpdateTime) * rewardRate * PRECISION) /
+			_totalSupply;
+	}
+
+	function _getRewardPerTokenPerUser(address _account)
+		internal
+		view
+		returns (uint256)
+	{
+		return
+			rewardsPerTokenAt[lastEpochIndex] -
+			rewardsPerTokenAt[stakersInfo[_account].indexOfLastEpochStaked + 1];
+	}
+
+	function earned(address account) public view  returns (uint256) {
+		return
+			(stakersInfo[account].balance *
+				(rewardPerToken() -
+					stakersInfo[account].rewardPerTokenPaid)) /
+			PRECISION +
+			stakersInfo[account].reward;
+	}
+
+	function _updateReward(address _account) internal virtual {
+		lastUpdateTime = lastTimeRewardApplicable();
+		if (_account != address(0)) {
+			_addPendingStakesToBalanceOnTimeUpdate(_account);
+			stakersInfo[_account].reward = earned(_account);
+			stakersInfo[_account]
+					.rewardPerTokenPaid = _getRewardPerTokenPerUser(_account);
```

Comment:
how many time do i have to repeat myself that there is no rewardPerToken only indexes

Reply:
Then how can we keep track of how much we already paid to the user?